src/lib/config.ts

```diff
@@ -0,0 +1,417 @@
+ * therefore it would have a defined endpoint.
+ *
+ * Flags like --host, --port, etc., provided by the CLI take precedence over what's
+ * stored in ~/.fauna-shell.
+ *
+ * The --endpoint flag overries the default endpoint from fauna-shell.
+ *
+ * If ~/.fauna-shell doesn't exist, tries to build the connection options from the
+ * flags passed to the script.
+ *
+ * It always expect a secret key to be set in ~/.fauna-shell or provided via CLI
+ * arguments.
+ */
+export const lookupEndpoint = (flags: any, scope: string, role: string) => {
+  const rootConfig = ini.parse(readFileOpt(getRootConfigFile()));
+  const localConfigPath = getLocalConfigFile();
+  const localConfig = localConfigPath
+    ? ini.parse(fs.readFileSync(localConfigPath, "utf8"))
+    : undefined;
+
+  return lookupEndpoint0({
+    flags,
+    rootConfig,
+    localConfig,
+    scope,
+    role,
+  });
+};
+
+export const lookupEndpoint0 = (opts: {
+  flags?: { [key: string]: any };
+  rootConfig?: { [key: string]: any };
+  localConfig?: { [key: string]: any };
+  scope?: string;
+  role?: string;
+}): {
+  secret: string;
+  url: string;
+  graphqlHost: string;
+  graphqlPort: number;
+} => {
+  const flags = new Config("flag", opts.flags ?? {});
+
+  const rootConfig = new RootConfig(
+    new Config("config key", opts.rootConfig ?? {})
+  );
+  const localConfig = opts.localConfig
+    ? new LocalConfig(new Config("config key", opts.localConfig))
+    : undefined;
+
+  localConfig?.validate(rootConfig);
+
+  const urlFlag = getURLFromConfig(flags);
+  if (urlFlag !== undefined) {
+    try {
+      new URL(urlFlag);
+    } catch (e) {
+      throw new Error(`Invalid database URL: ${urlFlag}`);
+    }
+  }
+
+  let stack: Stack | undefined;
+
+  if (localConfig !== undefined) {
+    // If a local config is present, a stack and endpoint may be chosen. The
+    // stack may come from the `default` key in the local config, or the
+    // `--stack` flag.
+    const stackName = flags.strOpt("stack") ?? localConfig.defaultStack;
+
+    stack = stackName ? localConfig.stacks[stackName] : undefined;
+    if (stackName !== undefined && stack === undefined) {
+      throw new Error(`No such stack '${stackName}'`);
+    }
+  } else {
+    const stackName = flags.strOpt("stack");
+    if (stackName !== undefined) {
+      throw new Error(
+        `No .fauna-project was found, so stack '${stackName}' cannot be used`
+      );
+    }
+  }
+
+  // An endpoint must be chosen as well. The endpoint may come from (in order)
+  // the `--endpoint` flag, the `endpoint` key in the local config, or the
+  // `default` key in the root config.
+  const endpointName =
+    flags.strOpt("endpoint") ?? stack?.endpoint ?? rootConfig.defaultEndpoint;
+
+  const secretFlag = flags.strOpt("secret");
+  if (endpointName === undefined && secretFlag === undefined) {
+    throw new Error(
+      "No endpoint or secret set. Set an endpoint in ~/.fauna-shell, .fauna-project, or pass --endpoint"
```

Comment:
would passing `--secret` also suffice here?

Reply:
Yeah, its a bit hard to pack all that into one error. The start says `No endpoint or secret` set, and I'd like to encourage people to use endpoints instead of passing `--secret`.